### PR TITLE
Prioritize system catalog for scalar functions

### PR DESCRIPTION
**Pull request by peterboncz**
It is currently possible to alter the SQL semantics by creating a MACRO with the same name as an existing scalar (or aggregation) function.

MACRO's get registered in the specified database/catalog and schema (or in the default one). Putting them in the system catalog is prohibited already.

While stopping short of prohibiting defining MACROs with standard names, this PR makes sure that the standard function is used whenever in SQL the function is not explicitly qualified (by database and/or schema name); by prioritizing the default system catalog for lookups of scalar functions.

A side-effect of this measure is that (extension) catalogs will get fewer lookups, since every scalar/aggregation in a SQL query function prior to this PR was looked up first in all catalogs (and default system catalog, where they reside, typically last).

(in MotherDuck, this lookup behavior then could lead to opening transactions that require RPCs even if the query itself were purely local)

### CODE REVIEW

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -586,7 +608,8 @@ CatalogEntryLookup Catalog::TryLookupEntry(ClientContext &context, CatalogType t
 	reference_set_t<SchemaCatalogEntry> schemas;
 	if (IsInvalidSchema(schema)) {
 		// try all schemas for this catalog
-		auto entries = GetCatalogEntries(context, GetName(), INVALID_SCHEMA);
+		auto scalar_function_lookup = (type == CatalogType::SCALAR_FUNCTION_ENTRY);
```
Comment (Mytherin) [inline src/catalog/catalog.cpp:611]: I would argue this should be consistent for all function types, i.e.:

```
TABLE_FUNCTION_ENTRY
SCALAR_FUNCTION_ENTRY
AGGREGATE_FUNCTION_ENTRY
PRAGMA_FUNCTION_ENTRY
COPY_FUNCTION_ENTRY
```

File: src/catalog/catalog.cpp [hunk 2]
```diff
@@ -369,19 +369,24 @@ SimilarCatalogEntry Catalog::SimilarEntryInSchemas(ClientContext &context, const
 	return result;
 }
 
-vector<CatalogSearchEntry> GetCatalogEntries(ClientContext &context, const string &catalog, const string &schema) {
+vector<CatalogSearchEntry> GetCatalogEntries(ClientContext &context, const string &catalog, const string &schema,
+                                             bool scalar_function_lookup = false) {
```
Comment (Mytherin) [inline src/catalog/catalog.cpp:373]: Maybe just `function_lookup`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! I've thought about it some more and I think this is a good solution/idea. Let's pull this through to all function types instead of only scalar functions though.

**Comment by peterboncz** [comment]
there is a python performance regression failing. I do think this is a flake.

**Comment by Mytherin** [comment]
Superseded by https://github.com/duckdb/duckdb/pull/11151 which accomplishes the same goal, but without breaking backwards compatibility

**Pull request closed without merging**